akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala

```diff
@@ -116,81 +116,182 @@ private[akka] object Shard {
    * remembered entity ids
    */
-  case object RememberedButNotCreated extends EntityState
+  // FIXME: since remember entities on start has a hop via region this could be a (small) resource leak
+  // if the shard extractor has changed, the entities will stay in the map forever as RememberedButNotCreated
+  // do we really need to track it?
+  case object RememberedButNotCreated extends EntityState {
+    override def transition(newState: EntityState): EntityState = newState match {
+      case NoState                       => RememberedButNotCreated
+      case active: Active                => active
+      case remembering: RememberingStart => remembering
+      case _                             => throw new IllegalArgumentException(s"Transition from $this to $newState not allowed")
+    }
+  }
 
   /**
    * When remember entities is enabled an entity is in this state while
-   * its existence is being recorded in the remember entities store.
+   * its existence is being recorded in the remember entities store, or while the stop is queued up
+   * to be stored in the next batch.
    */
-  case object Remembering extends EntityState
-  sealed trait WithRef extends EntityState {
-    def ref: ActorRef
+  case class RememberingStart(ackTo: Option[ActorRef]) extends EntityState {
+    override def transition(newState: EntityState): EntityState = newState match {
+      case active: Active      => active
+      case r: RememberingStart => r // FIXME is this really fine, will potentially replace ackTo
+      case _                   => throw new IllegalArgumentException(s"Transition from $this to $newState not allowed")
+    }
   }
-  final case class Active(ref: ActorRef) extends WithRef
-  final case class Passivating(ref: ActorRef) extends WithRef
-  case object WaitingForRestart extends EntityState
+
+  sealed trait StopReason
+  final case object Passivating extends StopReason
+  final case object StartedElsewhere extends StopReason
 
   /**
-   * Entity has been stopped. If the entity was pasivating it'll be removed from state
-   * once the stop has been recorded.
+   * When remember entities is enabled an entity is in this state while
+   * its stop is being recorded in the remember entities store, or while the stop is queued up
+   * to be stored in the next batch.
    */
-  case object Stopped extends EntityState
+  case class RememberingStop(stopReason: StopReason) extends EntityState {
+    override def transition(newState: EntityState): EntityState = newState match {
+      case NoState => NoState
+      case _       => throw new IllegalArgumentException(s"Transition from $this to $newState not allowed")
+    }
+  }
 
-  final case class EntityTerminated(id: EntityId, ref: ActorRef)
+  sealed trait WithRef extends EntityState {
+    def ref: ActorRef
+  }
+  final case class Active(ref: ActorRef) extends WithRef {
+    override def transition(newState: EntityState): EntityState = newState match {
```

Comment:
It could be nice if this is SxE -> S rather than S -> S. Then the lines on the graph are also represented. I think this is good for now though!